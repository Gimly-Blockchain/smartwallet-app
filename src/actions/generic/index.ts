import { Linking } from 'react-native'
import SplashScreen from 'react-native-splash-screen'
import I18n from 'src/locales/i18n'
import { routeList } from 'src/routeList'
import { navigationActions, accountActions } from 'src/actions'
import { ThunkAction } from 'src/store'
import settingKeys from 'src/ui/settings/settingKeys'
import { withLoading, withErrorScreen } from '../modifiers'
import {
  AppWrapConfig,
  APPWRAP_UPDATE_CONFIG,
  APPWRAP_SHOW_LOADER,
  APPWRAP_REGISTER_CONFIG,
  APPWRAP_UNREGISTER_CONFIG,
} from 'src/reducers/generic'
import { AnyAction } from 'redux'

// Default delay on the loading state value before it can switch back to 'false'
const DEFAULT_LOADING_LATCH_DELAY_MS = 300
import { AppError, ErrorCode } from '@jolocom/sdk/js/src/lib/errors'
import { AsyncStorage } from 'react-native'
import crypto from 'crypto'
import { termsOfServiceDE } from 'src/ui/termsAndPrivacy/legalTexts'

export const showErrorScreen = (
  error: AppError | Error,
): ThunkAction => dispatch => {
  const appError: AppError =
    error.constructor === AppError
      ? (error as AppError)
      : new AppError(ErrorCode.Unknown, error)

  return dispatch(
    navigationActions.navigate({
      routeName: routeList.Exception,
      params: {
        returnTo: appError.navigateTo,
        error: appError,
      },
    }),
  )
}

export const initApp: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  try {
    const storedSettings = await backendMiddleware.storageLib.get.settingsObject()

    // locale setup
    /** locale setup
     * @dev Until German and Dutch terms are polished, only English is used.
     * previous code:
     */
    if (storedSettings.locale) I18n.locale = storedSettings.locale
    else storedSettings.locale = I18n.locale

    storedSettings.locale = I18n.locale

    await dispatch(loadSettings(storedSettings))

    const ret = await dispatch(accountActions.checkIdentityExists)

    // FIXME what happens if no identity and this is a deeplink?
    // navigationActions.handleDeepLink throws NoWallet
    // need to improve this UX

    // FIXME: get rid of these after setting up deepLinking properly using
    // react-navigation
    const handleDeepLink = (url: string) =>
      dispatch(
        withLoading(withErrorScreen(navigationActions.handleDeepLink(url))),
      )
    Linking.addEventListener('url', event => handleDeepLink(event.url))
    await Linking.getInitialURL().then(url => {
      if (url) handleDeepLink(url)
    })

    return ret
  } catch (e) {
    return dispatch(
      showErrorScreen(
        new AppError(ErrorCode.WalletInitFailed, e, routeList.Landing),
      ),
    )
  } finally {
    SplashScreen.hide()
  }
}

const hashString = (text: string) => {
  return crypto
    .createHash('sha256')
    .update(text)
    .digest('hex')
}

const TERMS_OF_CONDITIONS_KEY = 'TERMS_OF_CONDITIONS'
export const checkTermsOfService = (
  route: routeList,
  onSubmit?: () => void,
): ThunkAction => async dispatch => {
  const storageHash = await AsyncStorage.getItem(TERMS_OF_CONDITIONS_KEY)
  const currentHash = hashString(termsOfServiceDE)
  const shouldShowTerms = storageHash !== currentHash

  if (!shouldShowTerms && onSubmit) onSubmit()
  return dispatch(
    navigationActions.navigate({
      routeName: shouldShowTerms ? routeList.TermsOfServiceConsent : route,
      params: { nextRoute: route, onSubmit },
    }),
  )
}

export const storeTermsOfService = (
  route: routeList,
): ThunkAction => async dispatch => {
  const termsHash = hashString(termsOfServiceDE)
  await AsyncStorage.setItem(TERMS_OF_CONDITIONS_KEY, termsHash)

  dispatch(navigationActions.navigate({ routeName: route }))
}

export const loadSettings = (settings: { [key: string]: any }) => ({
  type: 'LOAD_SETTINGS',
  value: settings,
})

export const setLocale = (locale: string): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  await backendMiddleware.storageLib.store.setting(settingKeys.locale, locale)
  I18n.locale = locale
  dispatch({ type: 'SET_LOCALE', value: locale })

  // we need to reset the navigator so that all screens are re-rendered with the
  // new locale
  return dispatch(navigationActions.navigatorReset())
}

const setLoading = (value: boolean) => ({
  type: APPWRAP_SHOW_LOADER,
  value,
})

export const updateAppWrapConfig = (value: AppWrapConfig) => ({
  type: APPWRAP_UPDATE_CONFIG,
  value,
})
export const registerAppWrapConfig = (value: AppWrapConfig & {}) => ({
  type: APPWRAP_REGISTER_CONFIG,
  value,
})
export const unregisterAppWrapConfig = (value: AppWrapConfig) => ({
  type: APPWRAP_UNREGISTER_CONFIG,
  value,
})

export const showAppLoading = (
  shouldShow: boolean,
  latchDelay = DEFAULT_LOADING_LATCH_DELAY_MS,
): ThunkAction => async (dispatch, getState) => {
  const wasShowing = getState().generic.appWrapConfig.loading
  if (shouldShow && !wasShowing) {
    // for setting to "true" we do it immediately and set current state
    const ret = dispatch(setLoading(shouldShow))
    return new Promise<AnyAction>(resolve => {
      setTimeout(() => resolve(ret), latchDelay)
    })
  }

  if (wasShowing && !shouldShow) {
    // for setting to 'false', we delay by the latchDelay amount
    return dispatch(delayedSetLoading(shouldShow, latchDelay))
  }
}

let nextLoadingState = false
let timeout: number | undefined
const delayedSetLoading = (
  value: boolean,
  latchDelay: number,
): ThunkAction => async dispatch => {
  nextLoadingState = value
  if (timeout !== undefined) clearTimeout(timeout)
  timeout = setTimeout(() => {
    dispatch(setLoading(nextLoadingState))
    timeout = undefined
  }, latchDelay)
}
