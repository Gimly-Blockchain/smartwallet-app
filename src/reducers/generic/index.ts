import { AnyAction } from 'redux'
import { pickBy } from 'ramda'

export const APPWRAP_UPDATE_CONFIG = 'APPWRAP_UPDATE_CONFIG'
export const APPWRAP_SHOW_LOADER = 'APPWRAP_SHOW_LOADER'
export const APPWRAP_SET_LOCKED = 'APPWRAP_SET_LOCKED'
export const APPWRAP_REGISTER_CONFIG = 'APPWRAP_REGISTER_CONFIG'
export const APPWRAP_UNREGISTER_CONFIG = 'APPWRAP_UNREGISTER_CONFIG'

export interface AppWrapConfig {
  readonly withoutStatusBar: boolean
  readonly loading: boolean
  readonly dark: boolean
  readonly locked: boolean
}
const initialAppWrapAttrs: AppWrapConfig = {
  loading: false,
  locked: true,
  withoutStatusBar: false,
  dark: false,
}

export const isAppWrapConfigAttr = (key: string) => initialAppWrapAttrs[key] !== undefined
export const pickAppWrapConfigAttrs: (o: any) => AppWrapConfig =
  pickBy((v, k) => isAppWrapConfigAttr(k))

export interface AppWrapState {
  readonly appWrapConfig: AppWrapConfig
  readonly appWrapConfigsSet: AppWrapConfig[]
}
const initialAppWrapState: AppWrapState = {
  appWrapConfig: initialAppWrapAttrs,
  appWrapConfigsSet: []
}

// TODO(mnzaki): this code is unnecessarily complicated
// the complicated "ConfigsSet" keeping is to support nested <Wrapper>
// components
// - but we don't really need to support them
// - see 20APR[RN61.5] on notion.so:spaces/mnzaki

export const appWrapReducer = (
  state = initialAppWrapState,
  action: AnyAction,
): AppWrapState => {
  let appWrapConfig, appWrapConfigsSet
  switch (action.type) {
    case APPWRAP_REGISTER_CONFIG:
      const idx = state.appWrapConfigsSet.indexOf(action.value)
      if (idx > -1 || Object.keys(action.value).length === 0) return state
      appWrapConfig = {...state.appWrapConfig, ...action.value }
      appWrapConfigsSet = [...state.appWrapConfigsSet, action.value]
      return { appWrapConfig, appWrapConfigsSet }
    case APPWRAP_UNREGISTER_CONFIG:
      const idx2 = state.appWrapConfigsSet.indexOf(action.value)
      if (idx2 < 0) return state
      appWrapConfigsSet = state.appWrapConfigsSet.filter(s => s !== action.value)
      appWrapConfig = appWrapConfigsSet.reduce((prev, cur) => ({...prev, ...cur}), initialAppWrapAttrs)
      return { appWrapConfig, appWrapConfigsSet }

    case APPWRAP_SHOW_LOADER:
      return {...state, appWrapConfig: {...state.appWrapConfig, loading: action.value} }
    case APPWRAP_SET_LOCKED:
      return {...state, appWrapConfig: {...state.appWrapConfig, locked: action.value} }
    default:
      return state
  }
}
