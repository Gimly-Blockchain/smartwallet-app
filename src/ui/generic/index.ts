export { Exception } from './exception'
export { LoadingScreen } from './loading'
export { LoadingSpinner } from './loadingSpinner'
export { SlidingTabBar } from './slidingTabBar'
