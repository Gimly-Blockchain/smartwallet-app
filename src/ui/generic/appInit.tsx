import * as React from 'react'
import { connect } from 'react-redux'

import { ThunkDispatch } from 'src/store'
import { navigationActions, accountActions, genericActions } from 'src/actions'
import { Linking, Dimensions, Image, StyleSheet } from 'react-native'
import { withLoading, withErrorHandler } from 'src/actions/modifiers'
import { Container, CenteredText } from '../structure'
import { JolocomTheme } from 'src/styles/jolocom-theme'
import { AppError, ErrorCode } from 'src/lib/errors'
import { showErrorScreen } from 'src/actions/generic'
const image = require('src/resources/img/splashScreen.png')

interface Props extends ReturnType<typeof mapDispatchToProps> {}

const styles = StyleSheet.create({
  loadingContainer: {
    backgroundColor: '#05050d',
  },
  loadingText: {
    position: 'absolute',
    bottom: '5%',
    fontSize: 12,
    fontFamily: JolocomTheme.contentFontFamily,
    color: JolocomTheme.primaryColorSand,
    opacity: 0.7,
  },
})

export class AppInitContainer extends React.Component<Props> {
  constructor(props: Props) {
    super(props)
    this.props.doAppInit()
  }

  render() {
    const viewWidth: number = Dimensions.get('window').width
    const viewHeight: number = Dimensions.get('window').height

    return (
      <Container style={styles.loadingContainer}>
        <Image
          source={image}
          style={{
            bottom: '15%',
            width: viewWidth,
            height: viewHeight / 2,
          }}
        />
        <CenteredText style={styles.loadingText} msg={'POWERED BY JOLOCOM'} />
      </Container>
    )
  }
}

const mapDispatchToProps = (dispatch: ThunkDispatch) => ({
  doAppInit: async () => {
    const withErrorScreen = withErrorHandler(
      showErrorScreen,
      (err: Error) => new AppError(ErrorCode.AppInitFailed, err),
    )
    await dispatch(withErrorScreen(genericActions.initApp))
    await dispatch(withErrorScreen(accountActions.checkIdentityExists))
    const handleDeepLink = (url: string) =>
      dispatch(
        withLoading(withErrorScreen(navigationActions.handleDeepLink(url))),
      )

    // FIXME: get rid of these after setting up deepLinking properly using
    // react-navigation
    Linking.addEventListener('url', event => handleDeepLink(event.url))
    Linking.getInitialURL().then(url => {
      if (url) handleDeepLink(url)
    })
  },
})

export const AppInit = connect(
  null,
  mapDispatchToProps,
)(AppInitContainer)
