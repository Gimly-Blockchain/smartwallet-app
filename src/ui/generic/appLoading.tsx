import * as React from 'react'
import { connect } from 'react-redux'
import { RootState } from 'src/reducers'

import { LoadingScreen } from './loading'
import { LoadingSpinner } from './loadingSpinner'

interface Props extends ReturnType<typeof mapStateToProps> {}

export const AppLoadingContainer: React.SFC<Props> = props =>
  props.loading || props.loading2 ? (
    props.hasIdentity ? (
      <LoadingSpinner />
    ) : (
      <LoadingScreen />
    )
  ) : null

const mapStateToProps = (state: RootState) => ({
  loading: state.account.loading,
  loading2: state.sso.deepLinkLoading,
  hasIdentity: !!state.account.did.did,
})

export const AppLoading = connect(mapStateToProps)(AppLoadingContainer)
