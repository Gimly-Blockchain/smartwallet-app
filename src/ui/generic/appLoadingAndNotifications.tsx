import * as React from 'react'
import { connect } from 'react-redux'
import { RootState } from 'src/reducers'

import { LoadingSpinner } from './loadingSpinner'
import { Notifications } from '../notifications/containers/notifications'
import { Animated, StyleSheet } from 'react-native'
import { Colors } from 'src/styles'

interface Props extends ReturnType<typeof mapStateToProps> {}

const LOADING_SPINNER_FADE_DURATION = 150

const styles = StyleSheet.create({
  loadingContainer: {
    backgroundColor: Colors.white,
    position: 'absolute',
    // to cover things such as the qr code scanner
    zIndex: 1,
    flex: 1,
    width: '100%',
    height: '100%',
  },
})

const AppLoadingAndNotificationsContainer: React.FunctionComponent<
  Props
> = props => {
  const [loadingOpacity] = React.useState(new Animated.Value(0))
  const [showContainer, setShowContainer] = React.useState(false)
  const showRightNow = props.loading || showContainer

  const startAnimation = (shouldShow: boolean) => {
    setTimeout(() => {
      Animated.timing(loadingOpacity, {
        toValue: shouldShow ? 1 : 0,
        duration: LOADING_SPINNER_FADE_DURATION,
      }).start(done => {
        if (done.finished) setShowContainer(shouldShow)
      })
    })
  }

  if (props.loading && !showContainer) setShowContainer(true)
  if (props.loading != showContainer) startAnimation(props.loading)

  return props.loading || showRightNow ? (
    <Animated.View
      style={[
        showRightNow && styles.loadingContainer,
        { opacity: loadingOpacity },
      ]}
    >
      <LoadingSpinner />
    </Animated.View>
  ) : (
    <Notifications />
  )
}

const mapStateToProps = (state: RootState) => ({
  loading: state.generic.loading,
})

export const AppLoadingAndNotifications = connect(mapStateToProps)(
  AppLoadingAndNotificationsContainer,
)
