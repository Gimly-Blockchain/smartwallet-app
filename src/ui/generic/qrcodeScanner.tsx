import React from 'react'
import { StyleSheet, Text } from 'react-native'
import { connect } from 'react-redux'
import { Container } from 'src/ui/structure'
import { JolocomTheme } from 'src/styles/jolocom-theme'
import { Button } from 'react-native-material-ui'
import { QrScanEvent } from 'src/ui/generic/qrcodeScanner'
import { navigationActions } from 'src/actions'
import I18n from 'src/locales/i18n'
import { LoadingSpinner } from './loadingSpinner'
import strings from '../../locales/strings'
import { JolocomLib } from 'jolocom-lib'
import { interactionHandlers } from '../../lib/storage/interactionTokens'
import { ThunkDispatch } from '../../store'
import { showErrorScreen } from '../../actions/generic'
import { RootState } from '../../reducers'
import { withErrorHandling, withLoading } from '../../actions/modifiers'
import { NavigationScreenProps } from 'react-navigation'
import { AppError, ErrorCode } from '../../lib/errors'
import { setDeepLinkLoading } from '../../actions/sso'

const QRScanner = require('react-native-qrcode-scanner').default

export interface QrScanEvent {
  data: string
}

interface Props
  extends ReturnType<typeof mapStateToProps>,
    ReturnType<typeof mapDispatchToProps>,
    NavigationScreenProps {
}

interface State { }

const styles = StyleSheet.create({
  buttonText: {
    color: JolocomTheme.primaryColorBlack,
  },
})

export class QRcodeScanner extends React.Component<Props, State> {
  private scanner: typeof QRScanner
  private removeFocusListener: (() => void) | undefined

  constructor(props: Props) {
    super(props)
    if (this.props.navigation) {
      this.removeFocusListener = this.props.navigation.addListener('willFocus', () => this.scanner.reactivate()).remove
    }
  }

  componentWillUnmount() {
    if (this.removeFocusListener) this.removeFocusListener()
  }

  render() {
    const { loading, onScannerSuccess, onScannerCancel } = this.props
    return (
      <React.Fragment>
        {loading && <LoadingSpinner />}
        <Container>
          <QRScanner
            ref={(ref: React.Component) => this.scanner = ref}
            onRead={(e: QrScanEvent) => onScannerSuccess(e)}
            cameraStyle={{ overflow: 'hidden' }}
            topContent={<Text>{I18n.t(strings.YOU_CAN_SCAN_THE_QR_CODE_NOW)}</Text>}
            bottomContent={
              <Button
                onPress={onScannerCancel}
                style={{ text: styles.buttonText }}
                text={I18n.t(strings.CANCEL)}
              />
            }
          />
        </Container>
      </React.Fragment>
    )
  }
}

const mapStateToProps = ({
  account: {
    loading: { loading },
  },
}: RootState) => ({
  loading,
})

const mapDispatchToProps = (dispatch: ThunkDispatch) => ({
  onScannerSuccess: async (e: QrScanEvent) => {
    let interactionToken

    try {
      interactionToken = JolocomLib.parse.interactionToken.fromJWT(e.data)
    } catch (err) {
      return dispatch(showErrorScreen(new AppError(ErrorCode.ParseJWTFailed, err)))
    }

    const handler = interactionHandlers[interactionToken.interactionType]

    return handler
      ? dispatch(
        withLoading(setDeepLinkLoading)(
          withErrorHandling(showErrorScreen)(handler(interactionToken)),
        ),
      )
      : dispatch(showErrorScreen(new AppError(ErrorCode.Unknown, new Error('No handler found'))))
  },
  onScannerCancel: () => dispatch(navigationActions.goBack),
})

export const QRScannerContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(QRcodeScanner)
