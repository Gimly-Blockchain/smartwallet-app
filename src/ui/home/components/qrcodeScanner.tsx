import React from 'react'
import { Text, StyleSheet, Platform, BackHandler } from 'react-native'
import { Container } from 'src/ui/structure'
import { JolocomTheme } from 'src/styles/jolocom-theme'
import { Button } from 'react-native-material-ui'
import { QrScanEvent } from '../containers/types'
import I18n from 'src/locales/i18n';
const QRScanner = require('react-native-qrcode-scanner').default

interface Props {
  onScannerSuccess: (e: QrScanEvent) => void
  onScannerCancel: () => void
}

interface State { }

const styles = StyleSheet.create({
  buttonText: {
    color: JolocomTheme.primaryColorBlack
  }
})

export class QRcodeScanner extends React.Component<Props, State> {
  componentDidMount() {
    if (Platform.OS === 'android') {
      BackHandler.addEventListener('hardwareBackPress', () => {
        this.props.onScannerCancel()
      })
    }
  }

  componentWillUnmount() {
    BackHandler.removeEventListener('hardwareBackPress', this.props.onScannerCancel)
  }

  render() {
    const { onScannerSuccess, onScannerCancel } = this.props
    return (
      <Container>
        <QRScanner
          onRead={(e: QrScanEvent) => onScannerSuccess(e)}
          topContent={<Text>{ I18n.t('You can scan the qr code now!') }</Text>}
          bottomContent={(
            <Button
              onPress={() => onScannerCancel()}
              style={{ text: styles.buttonText }}
              text={ I18n.t("Cancel")}
            />
          )}
        />
      </Container>
    )
  }
}
