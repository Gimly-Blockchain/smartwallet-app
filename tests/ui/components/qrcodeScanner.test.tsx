import React from 'react'
import { QRcodeScanner } from 'src/ui/generic/qrcodeScanner'
import { shallow } from 'enzyme'

describe('QRCodeScanner component', () => {
  it('matches the snapshot with back handler', () => {
    const props: QRcodeScanner['props'] = {
      onScannerSuccess: jest.fn(),
      onScannerCancel: jest.fn(),
      loading: false,
      // @ts-ignore navigation param not needed
      navigation: null,
    }

    const rendered = shallow(<QRcodeScanner {...props} />)
    expect(rendered).toMatchSnapshot()
  })
})
