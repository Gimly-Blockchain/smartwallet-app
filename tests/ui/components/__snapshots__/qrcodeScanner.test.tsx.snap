// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QRCodeScanner component matches the snapshot with back handler 1`] = `
<Fragment>
  <Container>
    <QRCodeScanner
      bottomContent={
        <ThemedComponent
          onPress={[MockFunction]}
          style={
            Object {
              "text": Object {
                "color": "black",
              },
            }
          }
          text="Cancel"
        />
      }
      cameraProps={Object {}}
      cameraStyle={
        Object {
          "overflow": "hidden",
        }
      }
      cameraType="back"
      checkAndroid6Permissions={false}
      fadeIn={true}
      notAuthorizedView={
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 16,
                "textAlign": "center",
              }
            }
          >
            Camera not authorized
          </Text>
        </View>
      }
      onRead={[Function]}
      pendingAuthorizationView={
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 16,
                "textAlign": "center",
              }
            }
          >
            ...
          </Text>
        </View>
      }
      permissionDialogMessage="Need camera permission"
      permissionDialogTitle="Info"
      reactivate={false}
      reactivateTimeout={0}
      showMarker={false}
      topContent={
        <Text>
          You can scan the qr code now!
        </Text>
      }
      vibrate={true}
    />
  </Container>
</Fragment>
`;
